### PR TITLE
BGP unnumbered support for Cumulus

### PR DESCRIPTION
**Pull request by charlyforot**
* Add CUMULUS-BGPUN-MIB provided by Cumulus
* Add  "bgpPeerIface" column to the BgpPeers table (to know which interface is used for the BGP link)
* Use CUMULUS-BGPUN-MIB in bgp-peers module for discovery
* Use CUMULUS-BGPUN-MIB in bgp-peers module for polling
* Add BGP unumbered tests with CUMULUS-BGPUN-MIB

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Review by PipoCanaja — CHANGES_REQUESTED** [review]

File: sql-schema/282.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `bgpPeers` ADD `bgpPeerIface` VARCHAR(255) NOT NULL DEFAULT '' AFTER `bgpPeerLastErrorText`;
```
[the remark this answers is not in the document]
Comment (charlyforot) [reply sql-schema/282.sql:1]: Thank you, I'm going to correct this.

**Comment by PipoCanaja** [comment]
Hi @charlyforot 
Probably a silly question, but still : 
Why did you create a new DB column instead of using existing one ? Seems that Cumulus uses a different name "bgpPeerIface"  where all others are doing "bgpPeerRemoteAddr". 
Would it be a problem to use bgpPeerRemoteAddr ?

**Comment by charlyforot** [comment]
> Hi @charlyforot Probably a silly question, but still : Why did you create a new DB column instead of using existing one ? Seems that Cumulus uses a different name "bgpPeerIface" where all others are doing "bgpPeerRemoteAddr". Would it be a problem to use bgpPeerRemoteAddr ?

Hi @PipoCanaja, this is not a silly question, I checked the standard [BGP4-MIB](https://www.circitor.fr/Mibs/Html/B/BGP4-MIB.php) and the [CUMULUS-BGPUN-MIB](https://docs.nvidia.com/networking-ethernet-software/mibs/CUMULUS-BGPUN-MIB.txt) to understand the differences.
There are only two new fields in the BgpPeerEntry sequence (bgpPeerIface and bgpPeerDesc), the rest is exactly the same.

bgpPeerIface description : 

> "For BGP unnumbered neighbors routes are using interfaces as nexthop interfaces. This is the inteface from which nexthop route"

I noticed that bgpPeerIface contains the name of the interface for BGP unumbered links but it still used for BGP numbered links, it contains the next hop IP address.

Contrary to bgpPeerAddr that must contain :

> "The remote IP address of this entry's BGP peer."

So this is necessarily an IP address.
Cumulus uses these two fields.

Tell me if you are agree with that or if you want process differently.

Concerning the failed tests, I don't understand why the error "duplicate column" appeared. Did I modify a file that I should not ?

**Comment by laf** [comment]
If we're going to store interface names then my personal opinion (and it would be good to get @murrant and @PipoCanaja  views on this) would be to only use that new column for interface names, if you see an IP in `bgpPeerIface` then store that in  `bgpPeerRemoteAddr` so it's standardised across the board.

Then also add in some a lookup for devices that don't return an interface name so we get an interface name for all bgp peers if we can: `select ifName from ports JOIN ipv4_addresses ON ports.port_id=ipv4_addresses.port_id WHERE ipv4_addresses.ipv4_address='1.2.3.4' AND ports.device_id=X;` as an example.

**Comment by charlyforot** [comment]
> If we're going to store interface names then my personal opinion (and it would be good to get @murrant and @PipoCanaja views on this) would be to only use that new column for interface names, if you see an IP in `bgpPeerIface` then store that in `bgpPeerRemoteAddr` so it's standardised across the board.
> 
> Then also add in some a lookup for devices that don't return an interface name so we get an interface name for all bgp peers if we can: `select ifName from ports JOIN ipv4_addresses ON ports.port_id=ipv4_addresses.port_id WHERE ipv4_addresses.ipv4_address='1.2.3.4' AND ports.device_id=X;` as an example.

Thank you for your opinion. I think it's a good idea, I'm waiting for a confirmation from @PipoCanaja to make changes.

**Comment by PipoCanaja** [comment]
> If we're going to store interface names then my personal opinion (and it would be good to get @murrant and @PipoCanaja views on this) would be to only use that new column for interface names, if you see an IP in `bgpPeerIface` then store that in `bgpPeerRemoteAddr` so it's standardised across the board.
> 
> Then also add in some a lookup for devices that don't return an interface name so we get an interface name for all bgp peers if we can: `select ifName from ports JOIN ipv4_addresses ON ports.port_id=ipv4_addresses.port_id WHERE ipv4_addresses.ipv4_address='1.2.3.4' AND ports.device_id=X;` as an example.

Hi guys,
Makes sense to keep same columns for same data as much as possible. So IP goes in existing `bgpPeerRemoteAddr`. For interfaces, I would even try, if possible, to store interface_id (pointing to an existing interface_id of lnms DB) instead of a name. More compact for storage, for searches, etc. Type of `bgpPeerIface` would be INT instead of CHAR(255) or VARCHAR.

**Comment by charlyforot** [comment]
Hello, ./lnms dev:check is OK on my side but some tests are skipped. Could we run tests again here ?

**Comment by charlyforot** [comment]
Can I have some help about the failling test ?
Maybe, adding the new "bgpPeerIface" column was not a good idea ?

**Comment by murrant** [comment]
@charlyforot I don't think the new field adds any additional data.  Looks like you are just using it to cache the result of a query of ipv4 tables.  Is that a fair analysis?

If so, then we should just do that query when presenting the data in the webui instead of adding extra work during polling.  If there are performance issues with that (there shouldn't be) then we can try to figure out how to optimize it.

**Comment by charlyforot** [comment]
@murrant That's right, no more data. The errors were in tests because the 'bgpPeerIface' was not in the .json of other OS. I've regenerate .json with the new column and that work !

I've no performance issues but I have searched another solutions to fill the bgpPeerIface column to optimize the program :
1 - I think fill the column elsewhere than in the polling module could lead to lack of understanding in the future because all columns are filled in the polling module.
2 - I tried to create a relationship between BgpPeer and Ipv4Address, BgpPeer and Ipv6Address. And another one with Ipv4Address and Port, Ipv6Address and Port to get the ifIndex. But it increases the number of query when searching for the ifIndex in the Port table.

So I kept my solution in the polling module and I add a solution to optimize the search of ifIndex when the bgpPeerIface is already filled by a manufacturer in SNMP.

At this moment, i've not display this column in the UI, because I don't want to overload the UI routing page. What do you think about that ?

**Comment by charlyforot** [comment]
@PipoCanaja Could you have a look to my new commits please ?

**Comment by charlyforot** [comment]
Hello @PipoCanaja, 
Could you have a look on my modification ?

**Comment by murrant** [comment]
LGTM, but re-running tests.

**Comment by PipoCanaja** [comment]
Hi @charlyforot 
Fixing tests now. Will see how it goes.

**Comment by PipoCanaja** [comment]
save-test-data.php returns a file, but the test on that same file requires [f895716](https://github.com/librenms/librenms/pull/13785/commits/f895716daaf94ad90449b916010feaf058189b45) to be applied ... There is something me miss.

**Comment by PipoCanaja** [comment]
@murrant Tests are good now. But I can't explain why save-test-data generates a JSON that is later on refused by lnms dev:check (on the same host, same config etc) and in github tests as well...
Would it be the same issue https://github.com/librenms/librenms/pull/13487 is addressing ?

**Review by PipoCanaja — APPROVED** [review]
LGTM

**Comment by charlyforot** [comment]
Thanks for the review @PipoCanaja @murrant

**Comment by zarya** [comment]
It does not seem to work with a sort of default Cumulus 4.4.3 install the OID for LocalAs does not lookup a snmp walk of 1.3.6.1.2.1.15 does not turn up anything.

**Pull request merged**